File: src/AutomatonTheory/JavaLogic/Automaton.java
package AutomatonTheory.JavaLogic;

import java.util.ArrayList;
import java.util.List;

public abstract class Automaton {
    String AutomatonName = "";
    List<State> States = new ArrayList<State>();
    List<String> Alphabet = new ArrayList<String>();
    Automatons Type = Automatons.DFA;

    abstract boolean evaluateString(String stringEvaluate);
    abstract boolean addTransition(String originStateName, String destinyStateName, String symbol);

    //region States
    boolean addState(State newState){
        if(!existsState(newState.Name)){
            return States.add(newState);
        }
        return false;
    }

    boolean removeState(String stateName){
        State originState = getState(stateName);
        //borra las transiciones que salen de el
        originState.Transitions.removeAll(originState.Transitions);

        //borra trans que llegan a el
        for (State state : States){
            List<Transition> toRemoveTransitions = new ArrayList<Transition>();
            for(Transition trans : state.Transitions){
                if(trans.DestinyState.Name.equals(originState.Name)){
                    toRemoveTransitions.add(trans);
                }
            }
            state.Transitions.removeAll(toRemoveTransitions);
        }

        //se borra el mismo estado
        return States.remove(originState);
    }

    boolean existsState(String stateName){
        for(State state : States){
            if(state.Name.equals(stateName)){
                return true;
            }
        }
        return false;
    }

    boolean toggleAcceptanceState(String stateName){
        for(State state : States){
            if(state.Name.equals(stateName)){
                state.AcceptanceState = !state.AcceptanceState;
                return true;
            }
        }
        return false;
    }

    State getInitialState(){
        for(State state : States){
            if(state.InitialState){
                return state;
            }
        }
        return new State();
    }

    boolean setInitialState(String stateName){
        for(State state : States){
            if(state.Name.equals(stateName)){
                state.InitialState = true;
            }
            else{
                state.InitialState = false;
            }
        }
        return true;
    }

    State getState(String stateName){
        for(State state : States){
            if(state.Name.equals(stateName)){
                return state;
            }
        }
        return new State();
    }
    //endregion

    //region Transitions
    boolean removeTransition(String originStateName, String destinyStateName, String symbol){
        State originState = getState(originStateName);
        Transition transition = getTransition(originStateName, destinyStateName, symbol);
        if(transition != null){
            return originState.Transitions.remove(transition);
        }
        return false;
    }

    Transition getTransition(String originStateName,String destinyStateName, String symbol){
        for(State state : States){
            for(Transition transition : state.Transitions){
                if(state.Name.equals(originStateName) && transition.DestinyState.Name.equals(destinyStateName) && transition.Symbol.equals(symbol)){
                    return transition;
                }
            }
        }
        return null;
    }

    Boolean existsTransition(String originStateName, String destinyStatName, String symbol){
        for(State state : States){
            for(Transition transition : state.Transitions){
                if(state.Name.equals(originStateName) &&
                        transition.DestinyState.Name.equals(destinyStatName) &&
                        transition.Symbol.equals(symbol)){
                    return true;
                }
            }
        }
        return false;
    }
    //endregion

    String getAutomatonInfo(){
        String returnString  = "";
        returnString += "Alphabet:" + "\n";
        for(String symbol : Alphabet){
            returnString += "  * " + symbol + "\n";
        }
        returnString+= "\n" + "States: " + "\n";
        for(State state : States){
            returnString += " * " + state.Name + ", Initial State: " + state.InitialState + ", Acceptance State: "+ state.AcceptanceState + "\n";
        }
        returnString+= "\n" + "Transitions: " + "\n";
        for(State state : States){
            for(Transition transition : state.Transitions){
                returnString += " * Origin State: " + state.Name + ", Destiny State: " + transition.DestinyState.Name + ", Symbol: " + transition.Symbol + "\n";
            }
        }
        return returnString;
    }

    boolean setAlphabet(List<String> alphabet){
        Alphabet = alphabet;
        return true;
    }
}

File: src/AutomatonTheory/Kotlin/AutomatonLogic/Automaton.kt
package AutomatonTheory.Kotlin.AutomatonLogic

import java.util.*

public abstract class Automaton {
    public var AutomatonName = ""
    public var States: MutableList<State> = ArrayList()
    public var Alphabet: MutableList<String> = ArrayList()
    public var Type = Automatons.DFA

    public abstract fun evaluateString(stringEvaluate: String): Boolean
    public abstract fun addTransition(originStateName: String, destinyStateName: String, symbol: String): Boolean

    //region States
    public fun addState(newState: State): Boolean {
        if (!existsState(newState.Name)) {
            return States.add(newState)
        }
        return false
    }

    public fun removeState(stateName: String): Boolean {
        val originState = getState(stateName)
        //borra las transiciones que salen de el
        originState.Transitions.removeAll(originState.Transitions)

        //borra trans que llegan a el
        for (state in States) {
            val toRemoveTransitions = ArrayList<Transition>()
            for (trans in state.Transitions) {
                if (trans.DestinyState.Name == originState.Name) {
                    toRemoveTransitions.add(trans)
                }
            }
            state.Transitions.removeAll(toRemoveTransitions)
        }

        //se borra el mismo estado
        return States.remove(originState)
    }

    public fun existsState(stateName: String): Boolean {
        for (state in States) {
            if (state.Name == stateName) {
                return true
            }
        }
        return false
    }

    public fun toggleAcceptanceState(stateName: String): Boolean {
        for (state in States) {
            if (state.Name == stateName) {
                state.AcceptanceState = !state.AcceptanceState
                return true
            }
        }
        return false
    }

    public fun getInitialState() : State{
        for (state in States) {
            if (state.InitialState) {
                return state
            }
        }
        return State()
    }

    public fun setInitialState(stateName: String): Boolean {
        for (state in States) {
            if (state.Name == stateName) {
                state.InitialState = true
            } else {
                state.InitialState = false
            }
        }
        return true
    }

    public fun getState(stateName: String): State {
        for (state in States) {
            if (state.Name == stateName) {
                return state
            }
        }
        return State()
    }
    //endregion

    //region Transitions
    public fun removeTransition(originStateName: String, destinyStateName: String, symbol: String): Boolean {
        val originState = getState(originStateName)
        val transition = getTransition(originStateName, destinyStateName, symbol)
        if (transition != null) {
            return originState.Transitions.remove(transition)
        }
        return false
    }

    public fun getTransition(originStateName: String, destinyStateName: String, symbol: String): Transition? {
        for (state in States) {
            for (transition in state.Transitions) {
                if (state.Name == originStateName && transition.DestinyState.Name == destinyStateName && transition.Symbol == symbol) {
                    return transition
                }
            }
        }
        return null
    }

    public fun existsTransition(originStateName: String, destinyStatName: String, symbol: String): Boolean? {
        for (state in States) {
            for (transition in state.Transitions) {
                if (state.Name == originStateName &&
                        transition.DestinyState.Name == destinyStatName &&
                        transition.Symbol == symbol) {
                    return true
                }
            }
        }
        return false
    }
    //endregion

    public fun getAutomatonInfo() : String {
        var returnString = ""
        returnString += "Alphabet:" + "\n"
        for (symbol in Alphabet) {
            returnString += "  * " + symbol + "\n"
        }
        returnString += "\n" + "States: " + "\n"
        for (state in States) {
            returnString += " * " + state.Name + ", Initial State: " + state.InitialState + ", Acceptance State: " + state.AcceptanceState + "\n"
        }
        returnString += "\n" + "Transitions: " + "\n"
        for (state in States) {
            for (transition in state.Transitions) {
                returnString += " * Origin State: " + state.Name + ", Destiny State: " + transition.DestinyState.Name + ", Symbol: " + transition.Symbol + "\n"
            }
        }
        return returnString
    }

    public fun setAlphabet(alphabet: MutableList<String>): Boolean {
        Alphabet = alphabet
        return true
    }

    public fun getStatesNames() : Array<String> {
        val states = ArrayList<String>()
        for (state in States) {
            states.add(state.Name)
        }
        return states.toTypedArray()
    }

    public fun getAllAlphabet() : Array<String>{
        return Alphabet.toTypedArray()
    }
}
